### PR TITLE
Convert aligned block sync

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

File: third_party/nvfuser/csrc/lower_utils.cpp [hunk 1]
```diff
@@ -745,6 +745,28 @@ bool isScalarExpr(Expr* expr) {
...
+    return false;
+  }
+
+  auto inputs = InputsOf::output(val->fusion(), val);
+
+  for (auto input : inputs) {
+    if (auto ns = dynamic_cast<NamedScalar*>(input)) {
+      for (auto ptype : kParallelTypeTIDs) {
+        if (ns->name() == stringifyThread(ptype) ||
+            ns->name() == stringifyThreadSize(ptype)) {
```
Comment (naoyam) [inline third_party/nvfuser/csrc/lower_utils.cpp]: Do thread sizes matter?

File: third_party/nvfuser/csrc/lower2device.cpp [hunk 2]
```diff
@@ -132,6 +132,46 @@ class KIRCleaner : public OptOutDispatch {
...
+
+ private:
+  using kir::IrVisitor::handle;
+
+  void handle(kir::BlockSync* sync) final {
+    // Inspect all the scope expressions
+    for (auto expr : scope_exprs_) {
+      // All predicates we have today are can be thread dependent so avoid this
+      // in general.
+      if (expr->isA<kir::IfThenElse>()) {
```
Comment (naoyam) [inline third_party/nvfuser/csrc/lower2device.cpp]: Why can't you just use `dependsonthreadNamedScalars` for this case as well?

File: third_party/nvfuser/csrc/lower_utils.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_utils.cpp]: No, I don't think it matters. I removed this function completely and uses `getRegisterType` from expr simplifier to replace it.

File: third_party/nvfuser/csrc/lower2device.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower2device.cpp]: Good point, changed

File: third_party/nvfuser/csrc/expr_simplifier.cpp [hunk 3]
```diff
@@ -746,9 +748,17 @@ Val* unflattenRule(Val* value, const VarInfoMap& var_info) {
...
+        // ptr + int -> ptr and ptr - int -> ptr, so we use IrBuilder::addExpr
+        // and IrBuilder::subExpr here.
+        // Also, convert a + (-b) to a - b here for better readibility on
+        // generated code
+        auto uop = dynamic_cast<UnaryOp*>(rhs->definition());
+        if (uop != nullptr && uop->getUnaryOpType() == UnaryOpType::Neg) {
+          lhs = IrBuilder::subExpr(lhs, uop->in());
+        } else {
+          lhs = IrBuilder::addExpr(lhs, rhs);
+        }
```
Comment (zasdfgbnm) [inline third_party/nvfuser/csrc/expr_simplifier.cpp:761]: Unrelated to this PR, just beautify the printed CUDA kernel.

File: third_party/nvfuser/csrc/expr_simplifier.cpp [hunk 4]
```diff
@@ -279,15 +281,11 @@ Val* recurseDown(Val* value, std::function<Val*(Val*)> rule) {
...
-
-enum class RegisterType { GeneralPurpose, Uniform, Immediate, Unknown };
-
 RegisterType getRegisterType(Val* value, const VarInfoMap& var_info) {
   if (auto ns = dynamic_cast<NamedScalar*>(value)) {
-    if (ns->getParallelIndex().has_value()) {
+    if (ns->getParallelIndex() == ParallelType::TIDx ||
+        ns->getParallelIndex() == ParallelType::TIDy ||
+        ns->getParallelIndex() == ParallelType::TIDz) {
       return RegisterType::GeneralPurpose;
```
Comment (zasdfgbnm) [inline third_party/nvfuser/csrc/expr_simplifier.cpp:289]: I think this was a bug, and I fixed it here.

File: third_party/nvfuser/csrc/arith.h [hunk 5]
```diff
@@ -289,6 +289,9 @@ TORCH_CUDA_CU_API TensorView* log2(TensorView*);
 // neg
 TORCH_CUDA_CU_API Val* neg(Val*);
 TORCH_CUDA_CU_API TensorView* neg(TensorView*);
+// notOp
+TORCH_CUDA_CU_API Val* notOp(Val*);
```
Comment (zasdfgbnm) [inline third_party/nvfuser/csrc/arith.h:293]: `not` is a C++ keyword https://en.cppreference.com/w/cpp/keyword/not, so I have to use `notOp`

File: third_party/nvfuser/csrc/expr_simplifier.cpp [hunk 6]
```diff
@@ -1310,6 +1322,18 @@ Val* eliminateTrivialComputation(Val* value, const VarInfoMap& var_info) {
...
+    auto optype = uop->getUnaryOpType();
+    if (optype == UnaryOpType::Neg || optype == UnaryOpType::Not) {
+      auto uop_in = dynamic_cast<UnaryOp*>(uop->in()->definition());
+      if (uop_in != nullptr) {
+        auto optype_in = uop_in->getUnaryOpType();
+        if (optype == optype_in) {
+          return uop_in->in();
+        }
+      }
+    }
```
Comment (zasdfgbnm) [inline third_party/nvfuser/csrc/expr_simplifier.cpp:1336]: Unrelated, just feel good to have it.

File: third_party/nvfuser/test/test_gpu1.cpp [hunk 7]
```diff
@@ -1199,9 +1199,11 @@ TEST_F(NVFuserTest, FusionParser_CUDA) {
   // 2. use a fuzzy compare (ignore non-significant whitespaces for example)
   const std::string expected_kernel = R"(
 __global__ void CUDAGeneratedKernel(Tensor<float, 1> T0, Tensor<float, 1> T1, Tensor<float, 1> T3) {
+  int64_t i53;
+  i53 = ((nvfuser_index_t)blockIdx.x) * 128;
   int64_t i54;
-  i54 = (((nvfuser_index_t)blockIdx.x) * 128) + ((nvfuser_index_t)threadIdx.x);
-  if ((i54 < T0.size[0])) {
+  i54 = i53 + ((nvfuser_index_t)threadIdx.x);
+  if ((((nvfuser_index_t)threadIdx.x) < (T0.size[0] - i53))) {
```
Comment (zasdfgbnm) [inline third_party/nvfuser/test/test_gpu1.cpp:1206]: `blockIdx.x` is now moved to rhs of `<`

File: third_party/nvfuser/csrc/expr_simplifier.cpp [hunk 8]
```diff
@@ -1865,8 +1889,6 @@ Val* simplifyExpr(Val* value, const std::list<VarInfo>& variables) {
     RUN_PASS(eliminateTrivialComputation);
     RUN_PASS(eliminateTrivialPredicate);
     RUN_PASS(simplifyDivisibleDivMod);
-    RUN_PASS(eliminateTrivialPredicate);
-    RUN_PASS(simplifyDivisibleDivMod);
```
Comment (zasdfgbnm) [inline third_party/nvfuser/csrc/expr_simplifier.cpp:1869]: This looks like an error in resolving merge conflicts

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**